doc/debug_framework_design_spec.md

```diff
@@ -0,0 +1,394 @@
+
+**4. Helper Scripts**
+New utility scripts are provided to:  
+- Print "headline" information for a quick summary of the system state.
+- Facilitate upload of the collected information.
+- Enforce policy on the number of debug-related files.
+
+# 3 Design
+## 3.1 Overview
+### 3.1.1 Framework for dump routines
+
+**Block Diagram**
+![Debug Framework Block Diagram](https://github.com/anilsadineni/SONiC/blob/debug_framework_HLD/images/debug_framework_block_diagram.png)
+
+
+Debug framework provides an API for components to register with the framework. It also provides interface for administrator, assert utilities, error handling functions to invoke framework actions. Framework uses Redis notifications for communicating the trigger message from entity invoking the trigger to the registered components. Framework actions are configurable. These actions are applicable uniformly to all registered components.  
+
+Debug Framework itself will not have an executing context. Framework actions are performed in the registered component's thread context. Trigger APIs are executed in the context of invoking utility or invoking process.  
+Please refer [Flow Diagram](#4-flow-diagrams) for execution sequence.
+
+**Framework Responsibilities**
+
+Each registered component provides the routines to dump the component-specific state via callback functions. It is then the responsibility of the framework to 
+- Monitor for specific system triggers or listen on framework trigger APIs invocation
+- Process the trigger and call appropriate dump routines 
+- Manage the output location
+- Follow up with post actions like redirect summary to console/ compress the files and upload the file.
+
+**Component Responsibilities**
+
+Components implement dump routines following below guidelines
+- Implement a function of type DumpInfoFunc.  
+  `typedef void (*DumpInfoFunc)(std::string, KeyOpFieldsValuesTuple);`
+- Within the function, dump the information into a string buffer and use a macro `SWSS_DEBUG_PRINT`.  Macro will further process the information based on intended framework action. Wrap the function that uses multiple invocations of `SWSS_DEBUG_PRINT` with `SWSS_DEBUG_PRINT_BEGIN` and `SWSS_DEBUG_PRINT_END`.
+- Handle KeyOpFieldsValuesTuple as argument. Arguments like dumpType and pass thru arguments that instruct the level of info dump ( short summary/specific info/filtered output/full info) should be handled by components. These arguments are opaque to framework. Additional optional arguments like output location and post-actions are for the framework consumption.  
+- Interpreting arguments is at discretion of components.  Components may choose to dump same info for all trigger types ignoring input arguments.
+- Dump routines registered with framework should be thread safe. Necessary synchronization needs to be ensured.

```

Comment:
orchagent is single thread application, it has no lock. how are you going to make the dump routine to be thread safe?

Reply:
If any application is not thread safe we recommend to use option #2 to link with framework.  In our pilot implementation for dump routines of  we chose option #2 for orchagent